state/protocol/badger/mutator.go

```diff
@@ -508,6 +508,9 @@ func (m *FollowerState) insert(ctx context.Context, candidate *flow.Block, certi
 		return fmt.Errorf("could not retrieve block header for %x: %w", parentID, err)
 	}
 
+	// TODO: Move this outside of insert. Per method documentation:
+	//  > Caller is responsible for ensuring block validity.
+	//  > No errors are expected during normal operation.
```

Comment:
I agree we should move validation of protocol state evolution out of `insert` and into a separate method.

I suggest Option 1, but have the protocol state evolution method consolidate and return all updates as a list of `dbUpdates`, rather than:
> Then, we could have some container struct that comprises all the objects returned by build

```go
func (m *FollowerState) protocolStateExtend(...) (transaction.DeferredDBUpdate, error) { ... }
```

Reasoning: unless we refactor `Build` to return _all_ the objects to insert, we will anyway be dealing with deferred updates as badger transaction anonymous functions, at least in part. If we have the concept of deferred database updates passed as anonymous functions at all, I'd prefer to encode all of them that way. 

Regarding Option 2, I think this is beneficial only if the members of the `deferredPersist` struct are objects to be inserted, not anonymous badger functions. Like you said, otherwise it conceptually has the same problem as using the simpler `[]func(*badger.Transaction) error`, that database updates are being created at several disparate levels of the code. 

Overall, when the deferred database updates don't have dependencies (which is the case here, and often), the deferred update pattern works well.

Reply:
Currently `StateMutator`  prepares a bunch of deferred updates as well as logic in `insert`. To me, conceptually, it's the same logic but spread over different places to have more modularization instead of having one huge `insert` method. I totally see `insert` and `StateMutator` logic as almost equivalent, `ProtocolStateMutator` prepares it's set of updates(which is unknown for anybody else), `insert` prepares it's set of updates, the only difference is that `insert` executes all of them as well. 
Currently we have two modules that apply their own updates, we can imagine that there are multiple modules like `ProtocolStateMutator` the logical thing for me would be to visit each of them, give them a chance to generate their deferred updates and then just commit when everything has been collected. We could try to separate insert into 3 logical sub-routines:
1) prepare protocol state changes
2) prepare block tree changes
3) submit all changes to the DB, when this happens all subsequent steps should also succeed, if any step fails all of the modules should keep their invariant(as it is with dynamic protocol state, this is achieved mostly by creating mutator for each block and applying state changes by deferred db update). If we could utilize similar logic for other parts(only block tree at the moment) we could build an elegant solution using visitor approach.

Reply:
### Update: 

I have iterated a few times over the code and found that it quite challenging to find a solution that does not distribute some part of the logic over multiple functions. 
* We have the business logic for Extending the state, which we desire to proceed in some logical flow. This produces deferred data base updates and  deferred emission of notifications (to be emitted only if the database transaction succeeded)
*  I did not manage to find some design that both keeps the business logic clean and has all the data base updates consolidated in one place 😞 
* Lastly, I didn't like the usage of a slice for storing deferred database operations for two reasons:
  * appending to a slice will probably copy the data a few times (minor concern)
  * We are not only dealing with deferred database operations but also with deferred events and possibly metrics. Instead of separate slices, I think it is better to have one object that manages all the deferred operations. 

The approach I finally settled on is [`DeferredDbOps`](https://github.com/onflow/flow-go/blob/e263ae188fc79ece41dca560e03cf1ffb89c3984/storage/badger/transaction/deferred_update.go#L23-L55). In a nutshell, this is an extension of the [`transaction.Tx`](https://github.com/onflow/flow-go/blob/e263ae188fc79ece41dca560e03cf1ffb89c3984/storage/badger/transaction/tx.go#L29-L32) concept. Using  [`DeferredDbOps`](https://github.com/onflow/flow-go/blob/e263ae188fc79ece41dca560e03cf1ffb89c3984/storage/badger/transaction/deferred_update.go#L23-L55), you can iteratively add deferred data base operations and callbacks on success. Internally, it implements a stack-like structure (not using a slice to avoid repeated copy), which I think suites the one-time-use pattern best.  

Overall, I liked Yurii's proposed approach (comment right above). The current implementation is the closest I got to it without diving into parallelization right away. I feel the current pattern with  [`DeferredDbOps`](https://github.com/onflow/flow-go/blob/e263ae188fc79ece41dca560e03cf1ffb89c3984/storage/badger/transaction/deferred_update.go#L23-L55) would make it easier to add concurrency later, but I am not sure. 


Curious to hear you comments and feedback.